torch/csrc/jit/script/init.cpp

```diff
@@ -363,6 +346,37 @@ Resolver pythonResolver(ResolutionCallback rcb) {
   };
 }
 
+FunctionSchema getSchemaWithDefaults(
+    std::unordered_map<std::string, py::object>& default_args,
+    FunctionSchema& schema) {
+  std::vector<Argument> new_args;
+  for (auto& arg : schema.arguments) {
+    auto it = default_args.find(arg.name);
+    if (it != default_args.end()) {
+      IValue value = *constantToIValue(it->second);
```

Comment:
There is no need for `constantToIValue`, we already have `inline IValue toIValue(py::handle obj, const TypePtr& type);` which is the correct call for this purpose. Most importantly, this pathway ensure that the returned `IValue` is in fact the right `type`. The refactored `constantToIValue` does not ensure that so this code can cause the program to throw an exception.

Reply:
Without type hints the schema I was getting for the functions in `test_default_values_function` looked like this
```
simple_fn(Tensor x, Tensor a=<default>, Tensor b=<default>, Tensor c=<default>) ->
bool_fn(Tensor x, Tensor a=<default>, Tensor flag=<default>) ->
```
Which would make `inline IValue toIValue(py::handle obj, const TypePtr& type)` fail since it would try to make an `autograd::Variable` from a scalar.

I added a check for the correct type + accompanying test for the correct types in the latest commit